internal/app/registry.go

```diff
@@ -0,0 +1,60 @@
+package app
+
+import (
+	"errors"
+	"sync"
+
+	"github.com/asynkron/protoactor-go/actor"
+)
+
+var errAlreadyRegistered = errors.New("factory already registered")
+
+var (
+	registry Registry
+	rwMu     sync.RWMutex
+)
+
+type (
+	Registry    map[string]FactoryFunc
+	FactoryFunc func(any) actor.Producer
+)
+
+func init() {
+	registry = make(Registry)
+}
+
+// Register registers the given factory with the given name.
+// It returns an error if a factory with the same name is already registered.
+func Register(name string, factory FactoryFunc) error {
+	rwMu.Lock()
+	defer rwMu.Unlock()
+
+	if _, exists := registry[name]; exists {
+		return errAlreadyRegistered
+	}
+	registry[name] = factory
+
+	return nil
+}
+
+// RegisterMust registers the given factory with the given name.
+// It panics if a factory with the same name is already registered.
+func RegisterMust(name string, factory FactoryFunc) {
+	if err := Register(name, factory); err != nil {
+		panic(err)
+	}
+}
+
+// All returns a function that iterates over the registry with a callback.
+func All() func(yield func(string, FactoryFunc) bool) {
+	rwMu.RLock()
+	defer rwMu.RUnlock()
+
+	return func(yield func(string, FactoryFunc) bool) {
+		for k, v := range registry {
+			if !yield(k, v) {
+				break
+			}
+		}
+	}
+}
```

Comment:
1. You don't have to add a type alias. You could use [iter.Seq2[U,B]](https://pkg.go.dev/iter#Seq2) instead.

```go
func All() iter.Seq2[string, FactoryFunc] {
...
  return func(yield func(string, FactoryFunc) bool) {
    ...
  }
...
}
```

2. Oh, you prefer make a systematic copy of the registry each time you want to iterate over it, rather than maintain a read lock for a little longer? I don't think it's a good practice. I'd prefer using the read-lock. However, my implementation was incorrect because I placed the lock outside the iteration function, making it ineffective.

Here’s the revised function, which acts as an implementation of a "concurrent" map. I took the opportunity to rename it to `Seq` which I find more appropriate.

```go
// Seq returns a function that iterates over the registry with a callback.
func Seq() iter.Seq2[string, FactoryFunc] {
	return func(yield func(string, FactoryFunc) bool) {
		rwMu.RLock()
		defer rwMu.RUnlock()

		for k, v := range registry {
			if !yield(k, v) {
				return
			}
		}
	}
}
```